### ISSUE TITLE
Questions about neurons in SipkingSSMs

### ISSUE THREAD

**Issue by Precola** [shenshuaijie/SDN#1]
Hey!
After seeing the SS4D class, may I ask if only "y = self.dropout(self.neuron(y))" this place used spike trains?  May I ask the average firing rate during the train? I see the time step in your paper is very large e.g. 1024, 2048.. does it have a big effect on ocupying the memory of GPU?
`
    def forward(self, u, **kwargs):  # absorbs return_output and transformer src mask
        """Input and output shape (B, H, L)"""
        if not self.transposed:
            u = u.transpose(-1, -2)
        L = u.size(-1)

        # Compute SSM Kernel
        k, _ = self.kernel(L=L)  # (H L)
        if self.bidirectional:
            k0, k1 = rearrange(k, "(s c) h l -> s c h l", s=2)
            k = F.pad(k0, (0, L)) + F.pad(k1.flip(-1), (L, 0))

        # Convolution
        k_f = torch.fft.rfft(k, n=2 * L)  # (C H L)
        u_f = torch.fft.rfft(u, n=2 * L)  # (B H L)

        y = torch.einsum("bhl,chl->bchl", u_f, k_f)

        y = torch.fft.irfft(y, n=2 * L)[..., :L]  # (B C H L)

        # Compute D term in state space equation - essentially a skip connection
        y = y + torch.einsum("bhl,ch->bchl", u, self.D)

        y = rearrange(y, "b c h l -> b (c h) l")

        if self.learnable_vth:
            y = y / torch.exp(self.ln_vth)

        y = self.dropout(self.neuron(y))
        y = self.output_linear(y)
        if not self.transposed:
            y = y.transpose(-1, -2)
        return (
            y,
            None,
        )  
`

**Comment by shenshuaijie** [comment]
1. Correct, LIF neurons are exclusively incorporated after SSM blocks.
2. Currently, I haven’t yet conducted empirical measurements on firing rates during the training phase.
3. As a language model, SpikingSSMs' time-step is intrinsically aligned with token sequence length. Therefore, there is no significant additional memory consumption.
